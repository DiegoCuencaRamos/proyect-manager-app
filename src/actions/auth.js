import { auth, GoogleAuthProvider, FacebookAuthProvider, TwitterAuthProvider, GithubAuthProvider } from "../firebase"

// LOGIN
const login = (uid) => ({
    type: 'auth/login',
    payload: uid
})

// Login-middleware
const startGoogleLogin = () => {
    return (dispatch, getState) => {
        return auth.signInWithRedirect(GoogleAuthProvider)
            .catch(e => console.log(e.message))
    }
}

const startFacebookLogin = () => {
    return (dispatch, getState) => {
        return auth.signInWithRedirect(FacebookAuthProvider)
            .catch(e => console.log(e.message))
    }
}

const startTwitterLogin = () => {
    return (dispatch, getState) => {
        return auth.signInWithRedirect(TwitterAuthProvider)
            .catch(e => console.log(e.message))
    }
}

const startGithubLogin = () => {
    return (dispatch, getState) => {
        return auth.signInWithRedirect(GithubAuthProvider)
            .catch(e => console.log(e.message))
    }
}

const startEmailLogin = (email, password, setErrorMessage) => {
    return (dispath, getState) => {
        return auth.signInWithEmailAndPassword(email, password)
            .catch(e => {
                console.log(e.message)
                if(e.message === 'The password is invalid or the user does not have a password.') {
                    setErrorMessage('he password is invalid')
                } else {
                    setErrorMessage(e.message)
                }  
            })
    }
}

const startEmailSignUp = (email, password, setErrorMessage) => {
    return (dispatch, getState) => {
        return auth.createUserWithEmailAndPassword(email, password)
            .catch(e => {
                console.log(e.message)
                setErrorMessage(e.message)
            })
    }
}

// LOGOUT
const logout = () => ({
    type: 'auth/logout'
})

const startLogout = () => {
    return (dipatch, getState) => {
        return auth.signOut()
    }
}

export {
    login,
    startGoogleLogin,
    startFacebookLogin,
    startTwitterLogin,
    startGithubLogin,
    startEmailLogin,
    startEmailSignUp,
    logout,
    startLogout
}