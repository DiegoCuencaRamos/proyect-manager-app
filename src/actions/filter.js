const searchFilterChanged = (searchText = '') => ({
    type: 'filter/serachFilterChanged',
    payload: searchText
})

const sortByFilterChanged = (sortBy) => ({
    type: 'filter/sortByFilterChanged',
    payload: sortBy
})

const visualizeModeFilterChanged = (visualizeMode = 'list') => ({
    type: 'filter/visualizeModeFilterChanged',
    payload: visualizeMode
})

export { searchFilterChanged, sortByFilterChanged, visualizeModeFilterChanged }