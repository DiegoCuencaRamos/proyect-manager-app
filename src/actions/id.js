const setProyectId = id => ({
    type: 'id/setProyectId',
    payload: id
})

const setTaskId = id => ({
    type: 'id/setTaskId',
    payload: id
})

export { setProyectId, setTaskId }