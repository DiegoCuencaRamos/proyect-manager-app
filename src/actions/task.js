import database from '../firebase'

// 1. Add task
const addTask = (task) => ({
    type: 'task/addTask',
    payload: task
})

const startAddTask = (taskData = {}) => {
    return (dispatch, getState) => {
        const {
            proyectId = '', 
            name = '', 
            description = '', 
            status = ''
        } = taskData
        const task = { proyectId, name, description, status }

        return database.ref('/tasks').push(task)
            .then(ref => {
                dispatch(addTask({
                    id: ref.key,
                    ...task
                }))
            })     
    }
}

// 2. Edit task
const editTask = (id, updates) => ({
    type: 'task/editTask',
    payload: {
        id,
        updates
    }
})

const startEditTask = (id, updates) => {
    return (dispatch, getState) => {

        return database.ref(`/tasks/${id}`).update(updates)
            .then(() => {
                dispatch(editTask(id, updates))
            })
    }
}

// 3. Remove task
const removeTask = (id) => ({
    type: 'task/removeTask',
    payload: id
})

const startRemoveTask = (id) => {
    return (dispatch, getState) => {

        return database.ref(`/tasks/${id}`).remove()
            .then(() => {
                dispatch(removeTask(id))
            })
    }
}

// 4. Set tasks
const setTasks = (tasks) => ({
    type: 'task/setTasks',
    payload: tasks
})

const startSetTasks = () => {
    return (dispatch, getState) => {

        return database.ref(`/tasks`).once('value')
            .then(dataSnapshot => {
                const tasks = []

                dataSnapshot.forEach(dataChildSnapshot => {
                    tasks.push({
                        id: dataChildSnapshot.key,
                        ...dataChildSnapshot.val()
                    })
                })

                dispatch(setTasks(tasks))
            })
    }
}

export { 
    addTask, 
    startAddTask,
    editTask, 
    startEditTask,
    removeTask,
    startRemoveTask,
    setTasks,
    startSetTasks
}