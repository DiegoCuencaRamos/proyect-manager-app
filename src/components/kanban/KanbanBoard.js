import React from 'react'
import KanbanBoardHeader from './KanbanBoardHeader';
import KanbanBoardBody from './KanbanBoardBody';

const KanbanBoard = () => {
    // Variables
    const statuses =  ['todo', 'doing', 'done']
     
    // Render
    return (
        <section className="kanban">
            <div className="container">
                {statuses.map((status, index) => (
                    <div key={index} className={`kanban__board--${status}`}>
                        <KanbanBoardHeader status={status} />
                        <KanbanBoardBody status={status} />
                    </div> 
                ))}
            </div>
        </section>   
    )
}

export default KanbanBoard