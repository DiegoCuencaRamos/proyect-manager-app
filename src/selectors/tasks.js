const getVisibleTasks = (proyectId, tasks, { searchText, sortBy }) => {
    return tasks
        // Get filtered tasks by proyect
        .filter(task => task.proyectId === proyectId)
        // Get filtered tasks by searchText
        .filter(task => task.name.toLowerCase().includes(searchText.toLowerCase()))
        // Order tasks by criteria
        .sort((a, b) => {
            if(sortBy === 'name') {
                return a.name.toLowerCase() > b.name.toLowerCase() ? 1 : -1
            }
        }) 
}

export default getVisibleTasks