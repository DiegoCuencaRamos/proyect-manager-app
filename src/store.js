import { createStore } from 'redux'
import rootReducer from './reducer'

const enhancer = window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()

const store = createStore(
    rootReducer,
    /* preloadedState. */
    enhancer,
)

export default store