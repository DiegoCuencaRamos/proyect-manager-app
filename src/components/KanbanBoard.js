import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import { setProyectId } from '../actions/id'

const KanbanBoard = ({ isProyect }) => {
    // Variables
    const statuses =  ['todo', 'doing', 'done']
    const items = useSelector(state => state[ isProyect ? 'proyects' : 'tasks' ])
    // Events
    let onItemClicked
    if (isProyect) {
        const dispatch = useDispatch()
        onItemClicked = (e) => {
            const id = e.target.dataset.id
            dispatch(setProyectId(id))
        }
    }
    // Rendering
    return (
        <section className="kanban">
            <div className="container">
                {
                    statuses.map((status, index) => (
                        <div 
                            key={index}
                            className={`kanban__board--${status}`}
                        >
                            <div className="kanban__thead">
                                <h3>{status}</h3>
                            </div>
                            <div className="kanban__tbody">
                                <Link 
                                    key={index}
                                    to={isProyect ? `/add-proyect` : `/add-task`}
                                    className="kanban__item"
                                >
                                    <p className="kanban__add-plus">+</p>
                                    <p className="kanban__add-text">{`Create new ${isProyect ? 'proyect' : 'task'} here`}</p>
                                </Link>

                                    {
                                        items
                                            .filter(item => item.status === status)
                                            .map(({ id, name, description, invoice = undefined }) => (
                                                <Link 
                                                    key={id}
                                                    to={isProyect ? `/proyect/${id}` : `/task/${id}`}
                                                    className="kanban__item"
                                                    data-id={id}
                                                    onClick={onItemClicked}
                                                >
                                                    <p className="kanban__item__title">{name}</p>
                                                    <p className="kanban__item__description">{description}</p>
                                                    <p className="kanban__item__invoice">{invoice}</p>
                                                </Link>
                                            ))
                                    }
                            </div>
                        </div>
                    ))
                }
            </div>
        </section>
        
    )
}

export default KanbanBoard