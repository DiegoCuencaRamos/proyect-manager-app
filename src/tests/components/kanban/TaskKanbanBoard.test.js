import React from 'react'
import { mount } from 'enzyme'
import { MemoryRouter } from 'react-router'
import { Provider } from 'react-redux'
import store from '../../../store'
import KanbanBoard from '../../../components/kanban/KanbanBoard'
import { tasks } from '../../fixures/tasks'
import { addTask } from '../../../actions/task'

const renderWithWrapper = () => mount(
    <MemoryRouter initialEntries={[{ key: 'static' }]}>
        <Provider store={store}>
            <KanbanBoard />
        </Provider>
    </MemoryRouter>
)

// Render
test('Should render KanbanBoard correctly with no items', () => {
    const wrapper = renderWithWrapper()
    expect(wrapper).toMatchSnapshot()
})

test('Should render KanbanBoard correctly with tasks', () => {
    // Add tasks to the KanbanBoard
    tasks.forEach(task => {
        store.dispatch(addTask(task))
    })

    const wrapper = renderWithWrapper()
    expect(wrapper).toMatchSnapshot()
})

// Events
test('Should navigate to Add Proyect Page on add new item clicked', () => {
    const wrapper = renderWithWrapper()
    wrapper.find('.kanban__board--todo #kanbanNewItem Link').simulate('click', { button: 0 })
    expect(wrapper.find('Router').prop('history').location.pathname).toBe('/add-task')
})

test('Should navigate to Proyect Page proyect clicked', () => {
    const id = tasks[0].id
    const wrapper = renderWithWrapper()
    wrapper.find(`KanbanItem#${id} Link`).simulate('click', { button: 0 })
    expect(wrapper.find('Router').prop('history').location.pathname).toBe(`/task/${id}`)
})

