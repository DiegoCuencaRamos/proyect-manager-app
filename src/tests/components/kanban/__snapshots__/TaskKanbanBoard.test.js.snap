// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render KanbanBoard correctly with no items 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "static",
      },
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "static",
            "pathname": "/",
            "search": "",
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "static",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        }
      }
    >
      <KanbanBoard>
        <section
          className="kanban"
        >
          <div
            className="container"
          >
            <div
              className="kanban__board--todo"
              key="0"
            >
              <KanbanBoardHeader
                status="todo"
              >
                <div
                  className="kanban__thead"
                >
                  <h3>
                    todo
                  </h3>
                </div>
              </KanbanBoardHeader>
              <KanbanBoardBody
                status="todo"
              >
                <div
                  className="kanban__tbody"
                >
                  <div
                    className="kanban_item"
                    id="kanbanNewItem"
                  >
                    <Link
                      className="kanban__item"
                      to="/add-task"
                    >
                      <LinkAnchor
                        className="kanban__item"
                        href="/add-task"
                        navigate={[Function]}
                      >
                        <a
                          className="kanban__item"
                          href="/add-task"
                          onClick={[Function]}
                        >
                          <p
                            className="kanban__add-plus"
                          >
                            +
                          </p>
                          <p
                            className="kanban__add-text"
                          >
                            Create new task here
                          </p>
                        </a>
                      </LinkAnchor>
                    </Link>
                  </div>
                </div>
              </KanbanBoardBody>
            </div>
            <div
              className="kanban__board--doing"
              key="1"
            >
              <KanbanBoardHeader
                status="doing"
              >
                <div
                  className="kanban__thead"
                >
                  <h3>
                    doing
                  </h3>
                </div>
              </KanbanBoardHeader>
              <KanbanBoardBody
                status="doing"
              >
                <div
                  className="kanban__tbody"
                >
                  <div
                    className="kanban_item"
                    id="kanbanNewItem"
                  >
                    <Link
                      className="kanban__item"
                      to="/add-task"
                    >
                      <LinkAnchor
                        className="kanban__item"
                        href="/add-task"
                        navigate={[Function]}
                      >
                        <a
                          className="kanban__item"
                          href="/add-task"
                          onClick={[Function]}
                        >
                          <p
                            className="kanban__add-plus"
                          >
                            +
                          </p>
                          <p
                            className="kanban__add-text"
                          >
                            Create new task here
                          </p>
                        </a>
                      </LinkAnchor>
                    </Link>
                  </div>
                </div>
              </KanbanBoardBody>
            </div>
            <div
              className="kanban__board--done"
              key="2"
            >
              <KanbanBoardHeader
                status="done"
              >
                <div
                  className="kanban__thead"
                >
                  <h3>
                    done
                  </h3>
                </div>
              </KanbanBoardHeader>
              <KanbanBoardBody
                status="done"
              >
                <div
                  className="kanban__tbody"
                >
                  <div
                    className="kanban_item"
                    id="kanbanNewItem"
                  >
                    <Link
                      className="kanban__item"
                      to="/add-task"
                    >
                      <LinkAnchor
                        className="kanban__item"
                        href="/add-task"
                        navigate={[Function]}
                      >
                        <a
                          className="kanban__item"
                          href="/add-task"
                          onClick={[Function]}
                        >
                          <p
                            className="kanban__add-plus"
                          >
                            +
                          </p>
                          <p
                            className="kanban__add-text"
                          >
                            Create new task here
                          </p>
                        </a>
                      </LinkAnchor>
                    </Link>
                  </div>
                </div>
              </KanbanBoardBody>
            </div>
          </div>
        </section>
      </KanbanBoard>
    </Provider>
  </Router>
</MemoryRouter>
`;

exports[`Should render KanbanBoard correctly with tasks 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "static",
      },
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "static",
            "pathname": "/",
            "search": "",
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "static",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        }
      }
    >
      <KanbanBoard>
        <section
          className="kanban"
        >
          <div
            className="container"
          >
            <div
              className="kanban__board--todo"
              key="0"
            >
              <KanbanBoardHeader
                status="todo"
              >
                <div
                  className="kanban__thead"
                >
                  <h3>
                    todo
                  </h3>
                </div>
              </KanbanBoardHeader>
              <KanbanBoardBody
                status="todo"
              >
                <div
                  className="kanban__tbody"
                >
                  <div
                    className="kanban_item"
                    id="kanbanNewItem"
                  >
                    <Link
                      className="kanban__item"
                      to="/add-task"
                    >
                      <LinkAnchor
                        className="kanban__item"
                        href="/add-task"
                        navigate={[Function]}
                      >
                        <a
                          className="kanban__item"
                          href="/add-task"
                          onClick={[Function]}
                        >
                          <p
                            className="kanban__add-plus"
                          >
                            +
                          </p>
                          <p
                            className="kanban__add-text"
                          >
                            Create new task here
                          </p>
                        </a>
                      </LinkAnchor>
                    </Link>
                  </div>
                  <KanbanItem
                    color="green"
                    description="This is my task fixure 1"
                    endDate={1616108400000}
                    id="t1"
                    key="t1"
                    name="Task 1"
                    proyectId="p1"
                    startDate={1615762800000}
                    status="todo"
                  >
                    <div
                      className="kanban_item"
                    >
                      <Link
                        className="kanban__item"
                        data-id="t1"
                        to="/task/t1"
                      >
                        <LinkAnchor
                          className="kanban__item"
                          data-id="t1"
                          href="/task/t1"
                          navigate={[Function]}
                        >
                          <a
                            className="kanban__item"
                            data-id="t1"
                            href="/task/t1"
                            onClick={[Function]}
                          >
                            <p
                              className="kanban__item__title"
                            >
                              Task 1
                            </p>
                            <p
                              className="kanban__item__description"
                            >
                              This is my task fixure 1
                            </p>
                          </a>
                        </LinkAnchor>
                      </Link>
                    </div>
                  </KanbanItem>
                </div>
              </KanbanBoardBody>
            </div>
            <div
              className="kanban__board--doing"
              key="1"
            >
              <KanbanBoardHeader
                status="doing"
              >
                <div
                  className="kanban__thead"
                >
                  <h3>
                    doing
                  </h3>
                </div>
              </KanbanBoardHeader>
              <KanbanBoardBody
                status="doing"
              >
                <div
                  className="kanban__tbody"
                >
                  <div
                    className="kanban_item"
                    id="kanbanNewItem"
                  >
                    <Link
                      className="kanban__item"
                      to="/add-task"
                    >
                      <LinkAnchor
                        className="kanban__item"
                        href="/add-task"
                        navigate={[Function]}
                      >
                        <a
                          className="kanban__item"
                          href="/add-task"
                          onClick={[Function]}
                        >
                          <p
                            className="kanban__add-plus"
                          >
                            +
                          </p>
                          <p
                            className="kanban__add-text"
                          >
                            Create new task here
                          </p>
                        </a>
                      </LinkAnchor>
                    </Link>
                  </div>
                  <KanbanItem
                    color="brown"
                    description="This is my task fixure 2"
                    endDate={1616540400000}
                    id="t2"
                    key="t2"
                    name="Task 2"
                    proyectId="p1"
                    startDate={1616194800000}
                    status="doing"
                  >
                    <div
                      className="kanban_item"
                    >
                      <Link
                        className="kanban__item"
                        data-id="t2"
                        to="/task/t2"
                      >
                        <LinkAnchor
                          className="kanban__item"
                          data-id="t2"
                          href="/task/t2"
                          navigate={[Function]}
                        >
                          <a
                            className="kanban__item"
                            data-id="t2"
                            href="/task/t2"
                            onClick={[Function]}
                          >
                            <p
                              className="kanban__item__title"
                            >
                              Task 2
                            </p>
                            <p
                              className="kanban__item__description"
                            >
                              This is my task fixure 2
                            </p>
                          </a>
                        </LinkAnchor>
                      </Link>
                    </div>
                  </KanbanItem>
                </div>
              </KanbanBoardBody>
            </div>
            <div
              className="kanban__board--done"
              key="2"
            >
              <KanbanBoardHeader
                status="done"
              >
                <div
                  className="kanban__thead"
                >
                  <h3>
                    done
                  </h3>
                </div>
              </KanbanBoardHeader>
              <KanbanBoardBody
                status="done"
              >
                <div
                  className="kanban__tbody"
                >
                  <div
                    className="kanban_item"
                    id="kanbanNewItem"
                  >
                    <Link
                      className="kanban__item"
                      to="/add-task"
                    >
                      <LinkAnchor
                        className="kanban__item"
                        href="/add-task"
                        navigate={[Function]}
                      >
                        <a
                          className="kanban__item"
                          href="/add-task"
                          onClick={[Function]}
                        >
                          <p
                            className="kanban__add-plus"
                          >
                            +
                          </p>
                          <p
                            className="kanban__add-text"
                          >
                            Create new task here
                          </p>
                        </a>
                      </LinkAnchor>
                    </Link>
                  </div>
                  <KanbanItem
                    color="purple"
                    description="This is my task fixure 3"
                    endDate={1616968800000}
                    id="t3"
                    key="t3"
                    name="Task 3"
                    proyectId="p3"
                    startDate={1616626800000}
                    status="done"
                  >
                    <div
                      className="kanban_item"
                    >
                      <Link
                        className="kanban__item"
                        data-id="t3"
                        to="/task/t3"
                      >
                        <LinkAnchor
                          className="kanban__item"
                          data-id="t3"
                          href="/task/t3"
                          navigate={[Function]}
                        >
                          <a
                            className="kanban__item"
                            data-id="t3"
                            href="/task/t3"
                            onClick={[Function]}
                          >
                            <p
                              className="kanban__item__title"
                            >
                              Task 3
                            </p>
                            <p
                              className="kanban__item__description"
                            >
                              This is my task fixure 3
                            </p>
                          </a>
                        </LinkAnchor>
                      </Link>
                    </div>
                  </KanbanItem>
                </div>
              </KanbanBoardBody>
            </div>
          </div>
        </section>
      </KanbanBoard>
    </Provider>
  </Router>
</MemoryRouter>
`;
